lib/install.go

```diff
@@ -138,97 +136,78 @@ func install(product Product, tfversion string, binPath string, installPath stri
 	/* set symlink to desired version */
 	CreateSymlink(installFileVersionPath, binPath)
 	logger.Infof("Switched terraform to version %q", tfversion)
-	addRecent(product, tfversion, installPath) //add to recent file for faster lookup
+	addRecentVersion(product, tfversion, installPath) //add to recent file for faster lookup
 	return
 }
 
-// addRecent : add to recent file
-func addRecent(product Product, requestedVersion string, installPath string) {
+// getRecentFileName : Obtain recent version filename for product
+func getRecentFileName(product Product) string {
+	return recentFilePrefix + product.GetId() + ".json"
```

Comment:
To make user's experience smooth and to maintain backward compatibility, we probably need to add 1) a code to migrate old plain TXT file to JSON seamlessly to user and 2) a notice to let user know the migration is being done from plain TXT to JSON when TXT version of file exists and no JSON version exists yet.

Also are there any obstacles or reasoning that prevents us from using the same file for any Product?
I think it's more versatile to keep the single file to track recent versions of any product.

Reply:
> Also are there any obstacles or reasoning that prevents us from using the same file for any Product?
> I think it's more versatile to keep the single file to track recent versions of any product.

Sounds good

Reply:
I would suggest to keep the data in one file and use a dict/list combination to store it.
It could look like this
```
{
 "terraform": [
    "filename1",
    "filename2",
  ],
 "tofu": [
    "tofuname1",
    "tofuname2"
  ]
}
```

This would be a struct like this

```
type RecentFiles struct {
  Terraform []string `json:"terraform"`
  Tofu []string `json:"tofu"`
}
```

We could traverse/lookup easily through the lists and determine whether we have the local file or not.

Reply:
Thanks @MatrixCrawler,

Apart from having a "tigher" definition, is there any additional benefit over using a `map[string][]string`?

Because the pseudo code would look more like:

```
func getRecentVersions(product, file) []string {
  rawContent := readFile(file)
  recentVersionData := Unmarshal(rawContent)
  if product.Id in recentVersionData {
    return recentVersionData[product.Id]
  }
  return []
}
```

As opposed to:
```
func getRecentVersions(product, file) []string {
  rawContent := readFile(file)
  recentVersionData := Unmarshal(rawContent)
  if product.Id == "terraform" {
    return recentVersionData.Terraform
  }
  if product.Id == "opentofu" {
    return recentVersionData.OpenTofu
  }
  return []
}
```
which (I guess theoretically due to chance of this happening) would require up-keep to match the available products.

Though, I guess, to put the ownership of this back to the product itself, it could look more like:
```
func getRecentVersions(product, file) []string {
  rawContent := readFile(file)
  recentVersionData := Unmarshal(rawContent)
  return product.GetRecentVersionFromContent(recentVersionData)
}

func (p TerraformProduct) GetRecentVersionFromContent(recentVersionData RecentFiles) {
  return recentVersionData.Terraform
}
func (p OpenTofuProduct) GetRecentVersionFromContent(recentVersionData RecentFiles) {
  return recentVersionData.OpenTofu
}
```

Reply:
I am a fan of "clear" definitions. Sure, when you break it down this is nothing else than a map[string][]string but i think the code is better readable that way.
I have a suggestion for the conversion of the recent file to json
See https://github.com/warrensbox/terraform-switcher/pull/437 if you'd like.

Reply:
Thanks @MatrixCrawler

What do you think of the final suggestion in the previous comment, where we use a concrete structure, as you suggested, but the responsibility for returning the correct attribute is delegated to the product itself?

Reply:
Thanks @MatrixCrawler

What do you think of the final suggestion in the previous comment, where we use a concrete structure, as you suggested, but the responsibility for returning the correct attribute is delegated to the product itself?

Thanks for #437 - it looks how I'd have imagined - though seeing how you handled the data migration was good to see. 
Were you hoping to merge this into somewhere or purely as a demonstration? :)

Reply:
I did not want to override your suggestion. But if you are okay with that we could go with my solution. I need to add some tests though before we can merge that.

Reply:
> What do you think of the final suggestion in the previous comment, where we use a concrete structure, as you suggested, but the responsibility for returning the correct attribute is delegated to the product itself?
> 

That looks like a nice approach using an interface.
I went with a solution that simply returns exactly the same as the old functions did for simpicity. But i am not opposed to your solution at all.

Reply:
Can do - I had a local version using the new structure, but happy to go with yours :)

Reply:
Then i will add some more tests

Reply:
I'll just need to work on integrating those changes into #435, as this PR was based on the branch for that :)
I guess, adding the changes for OpenTofu might be a little.. odd, if #437 is merged before that.

Reply:
If you did want to merge your PR prior to #435 , then I think it makes sense, but if it does depend on it and/or makes more sense to ship together, is it worth either branching from #435 and making the changes or even branching from this PR to make the suggested changes?

Reply:
My branch is based on master an can be merged prior to #435. It should be fully backwards compatible.

Reply:
@yermulnik Just so you see the thread :)

Reply:
Thanks for highlighting it. It looks good to me (despite all those Golang bits, I have no understanding of, you guys are discussing here 😀)

Reply:
@MatrixCrawler Also the comments you added to #437 all make perfect sense (lib naming, ordering, pruning old items, so on) — thanks for taking thorough look 🤝

Reply:
Since yours closely resembles this PR and there's quite a few changes that will result in needing to be re-made for the #435 , which contains the "product" changes. I'll merge this PR (into #435) and I will deal with the fun merge of main :D)

Reply:
> and I will deal with the fun merge of main :D)

What an outstanding example of bravery =) Kudos to you man!

Reply:
IF this plan goes horrendously wrong and #437 causes turmoil - I can always back the merge :D